include/moveitem.h

```diff
@@ -189,57 +189,40 @@ inline int CountItem(const char* pcItemName, int usMin, int usInvSlots)
 /*********************************************************************************/
 // PackFind will fill a CItemLocation passed to it, or NULL if item not found
 //
-
-// finds an open slot in a bag that has items (not empty bags)
+// finds an open slot in a bag or inventory
 CItemLocation* PackFind(CItemLocation* pFreeSlot, ItemClient* pUnequipSlot)
 {
 	unsigned char  ucCurSize = 10;
-	unsigned short usSlot    = 0;
-
-	unsigned short usDecrement = 0;
+	ItemDefinition* pUnequipInfo = pUnequipSlot->GetItemDefinition();
 
-	// no user-specific slot given, find the first free bag slot
-	for (usSlot = InvSlot_FirstBagSlot; usSlot <= GetHighestAvailableBagSlot(); usSlot++)
+	for (unsigned short usSlot = InvSlot_FirstBagSlot; usSlot <= GetHighestAvailableBagSlot(); usSlot++)
 	{
 		if (ItemClient* pInvSlot = GetPcProfile()->GetInventorySlot(usSlot))
 		{
-			ItemDefinition* pItemInfo    = GetItemFromContents(pInvSlot);
-			ItemDefinition* pUnequipInfo = GetItemFromContents(pUnequipSlot);
+			ItemDefinition* pItemInfo = pInvSlot->GetItemDefinition();
 
 			if (pItemInfo && pUnequipInfo && pInvSlot->IsContainer()
 				&& pItemInfo->Combine != 2
 				&& pUnequipInfo->Size <= pItemInfo->SizeCapacity)
 			{
-				ItemClient* pLAST = pInvSlot->GetHeldItem(0);
-
-				unsigned short usPack = 0;
-				for (usPack = 0; pInvSlot->GetHeldItems().GetSize(); usPack++) {
+				for (unsigned short usPack = 0; usPack < pItemInfo->Slots; usPack++) {
 					if (!pInvSlot->GetHeldItem(usPack)) {
-						//DebugSpewAlways("no pointer at %hu of pack %hu", usPack, usSlot);
-						if (pLAST) {
-							pFreeSlot->InvSlot   = usSlot;
-							pFreeSlot->BagSlot   = usPack;
-							pFreeSlot->pBagSlot  = pInvSlot->GetHeldItem(usPack); // will be NULL
-							pFreeSlot->pInvSlot  = pInvSlot;
-							return pFreeSlot;
-						// else first bag slot was empty
-						} else {
-							usDecrement++;
-						}
-					} else if (usDecrement) {
-						//DebugSpewAlways("pack %hu minus decrement %hu equals %hu", usPack, usDecrement, usPack-usDecrement);
-						pFreeSlot->InvSlot    = usSlot;
-						pFreeSlot->BagSlot    = usPack;
-						pFreeSlot->pBagSlot   = pInvSlot->GetHeldItem(usPack - usDecrement); // will be NULL
-						pFreeSlot->pInvSlot   = pInvSlot;
+						pFreeSlot->InvSlot = usSlot;
+						pFreeSlot->BagSlot = usPack;
+						pFreeSlot->pBagSlot = pInvSlot->GetHeldItem(usPack);
+						pFreeSlot->pInvSlot = pInvSlot;
 						return pFreeSlot;
 					}
-					pLAST = pInvSlot->GetHeldItem(usPack);
 				}
-				usDecrement = 0;
-				pLAST = nullptr;
 			}
 		}
+		else {
+			pFreeSlot->InvSlot = usSlot;
+			pFreeSlot->BagSlot = INVALID_PACK;
+			pFreeSlot->pBagSlot = nullptr;
+			pFreeSlot->pInvSlot = nullptr;
+			return pFreeSlot;
+		}
```

Comment:
I don't think you should return an invalid slot on _any_ failure. You could still have valid options later in the loop, no?

Reply:
What do you mean by this?
This is the case where the slot is not a container and is empty, which means it can fit any item including the incoming one so its a valid slot to put the item.